src/crypto/internal/boring/openssl_port_rsa.c

```diff
@@ -88,49 +88,85 @@ int _goboringcrypto_RSA_sign_raw(GO_RSA *rsa, size_t *out_len, uint8_t *out, siz
 int _goboringcrypto_RSA_sign_pss_mgf1(RSA *rsa, size_t *out_len, uint8_t *out, size_t max_out,
                       const uint8_t *in, size_t in_len, const EVP_MD *md,
                       const EVP_MD *mgf1_md, int salt_len) {
-  if (in_len != EVP_MD_size(md)) {
-    return 0;
-  }
+	EVP_PKEY_CTX *ctx;
+	EVP_PKEY *pkey;
+	size_t siglen;
+
+	pkey = EVP_PKEY_new();
+	if (!pkey)
+		return 0;
+
+	if (EVP_PKEY_set1_RSA(pkey, rsa) <= 0)
+		return 0;
+	
+	ctx = EVP_PKEY_CTX_new(pkey, NULL /* no engine */);
+	if (!ctx)
+		return 0;
+
+	int ret = 0;
+
+	if (EVP_PKEY_sign_init(ctx) <= 0)
+		goto err;
+	if (EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PSS_PADDING) <= 0)
+		goto err;
+	if (EVP_PKEY_CTX_set_rsa_pss_saltlen(ctx, salt_len) <= 0)
+		goto err;
+	if (EVP_PKEY_CTX_set_signature_md(ctx, md) <= 0)
+		goto err;
+	
+	/* Determine buffer length */
+	if (EVP_PKEY_sign(ctx, NULL, &siglen, in, in_len) <= 0)
+		goto err;
+
+	if (max_out < siglen)
+		goto err;
+
+	if (EVP_PKEY_sign(ctx, out, &siglen, in, in_len) <= 0)
+		goto err;
+
+	*out_len = siglen;
+	ret = 1;
 
-  size_t padded_len = RSA_size(rsa);
-  uint8_t *padded = OPENSSL_malloc(padded_len);
-  if (padded == NULL) {
-    return 0;
-  }
+err:
+	EVP_PKEY_CTX_free(ctx);
 
-  int ret =
-      RSA_padding_add_PKCS1_PSS_mgf1(rsa, padded, in, md, mgf1_md, salt_len) &&
-      _goboringcrypto_RSA_sign_raw(rsa, out_len, out, max_out, padded, padded_len,
-                   RSA_NO_PADDING);
-  OPENSSL_free(padded);
-  return ret;
+	return ret;
 }
 
 int _goboringcrypto_RSA_verify_pss_mgf1(RSA *rsa, const uint8_t *msg, size_t msg_len,
                         const EVP_MD *md, const EVP_MD *mgf1_md, int salt_len,
                         const uint8_t *sig, size_t sig_len) {
-  if (msg_len != EVP_MD_size(md)) {
-    return 0;
-  }
-
-  size_t em_len = RSA_size(rsa);
-  uint8_t *em = OPENSSL_malloc(em_len);
-  if (em == NULL) {
-    return 0;
-  }
-
-  int ret = 0;
-  if (!_goboringcrypto_RSA_verify_raw(rsa, &em_len, em, em_len, sig, sig_len, RSA_NO_PADDING)) {
-    goto err;
-  }
-
-  if (em_len != RSA_size(rsa)) {
-    goto err;
-  }
-
-  ret = RSA_verify_PKCS1_PSS_mgf1(rsa, msg, md, mgf1_md, em, salt_len);
+	EVP_PKEY_CTX *ctx;
+	EVP_PKEY *pkey;
+
+	int ret = 0;
+
+	pkey = EVP_PKEY_new();
+	if (!pkey)
+		return 0;
+
+	if (EVP_PKEY_set1_RSA(pkey, rsa) <= 0)
+		return 0;
+	
+	ctx = EVP_PKEY_CTX_new(pkey, NULL /* no engine */);
+	if (!ctx)
+		return 0;
+
+	if (EVP_PKEY_verify_init(ctx) <= 0)
+		goto err;
+	if (EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PSS_PADDING) <= 0)
+		goto err;
+	if (EVP_PKEY_CTX_set_rsa_pss_saltlen(ctx, salt_len) <= 0)
+		goto err;
+	if (EVP_PKEY_CTX_set_signature_md(ctx, md) <= 0)
+		goto err;
```

Comment:
Same here as above.

Reply:
This still applies, although current callers probably use the same for md and mgf1_md.